opal/util/ethtool.c

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2016 Karol Mroz.  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "opal_config.h"
+
+#include <string.h>
+#include <limits.h>
+
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+#ifdef HAVE_NET_IF_H
+#include <net/if.h>
+#endif
+#ifdef HAVE_LINUX_ETHTOOL_H
+#include <linux/ethtool.h>
+#endif
+#ifdef HAVE_SYS_IOCTL_H
+#include <sys/ioctl.h>
+#endif
+#ifdef HAVE_LINUX_SOCKIOS_H
+#include <linux/sockios.h>
+#endif
+
+#include "opal/util/ethtool.h"
+#include "opal/util/if.h"
+
+#if !defined(HAVE_DECL_ETHTOOL_CMD_SPEED)
+static inline unsigned int
+ethtool_cmd_speed(const struct ethtool_cmd *ep)
+{
+    return (ep->speed_hi << 16) | ep->speed;
+}
+#endif
+
+/*
+ * Obtain an appropriate bandwidth for the interface if_name. On Linux, we
+ * get this via an ioctl(). Elsewhere or in the error case, we return the
+ * speed as 0.
+ */
+unsigned int
+opal_ethtool_get_speed (const char *if_name)
+{
+    unsigned int speed = 0;
+
+#if !defined(HAVE_DECL_SIOCETHTOOL) || !defined(HAVE_STRUCT_IFREQ) || !defined(HAVE_STRUCT_ETHTOOL_CMD)
```

Comment:
I'm sorry -- one last nit-pick, I swear... 😄 

I tend to prefer positive logic (vs. negative logic) -- I just find it easier to read/parse.  E.g.:

```c
#if defined(HAVE_DECL_SIOCETHTOOL) && defined(HAVE_STRUCT_IFREQ) && defined(HAVE_STRUCT_ETHTOOL_CMD)
    /*...long block with all the code ... */
#else
    return speed;
#endif
```